src/swap/do_action.rs

```diff
@@ -0,0 +1,212 @@
+use crate::{swap::db, SwapId};
+use comit::Timestamp;
+
+#[async_trait::async_trait]
+pub trait Do<E>
+where
+    Self: CheckMemory<E> + ShouldAbort + Execute<E> + Remember<E>,
+    E: Clone + Send + Sync + 'static,
+    <Self as Execute<E>>::Args: Send + Sync,
+{
+    async fn r#do(
+        &self,
+        beta_expiry: Timestamp,
+        execution_args: <Self as Execute<E>>::Args,
+    ) -> anyhow::Result<Next<E>> {
+        if let Some(event) = self.check_memory().await? {
+            return Ok(Next::Continue(event));
+        }
+
+        if self.should_abort(beta_expiry).await? {
+            return Ok(Next::Abort);
+        }
+
+        let event = Execute::<E>::execute(self, execution_args).await?;
+        self.remember(event.clone()).await?;
+
+        Ok(Next::Continue(event))
+    }
+}
+
+#[async_trait::async_trait]
+impl<E, A> Do<E> for A
+where
+    A: CheckMemory<E> + ShouldAbort + Execute<E> + Remember<E>,
+    E: Clone + Send + Sync + 'static,
+    <Self as Execute<E>>::Args: Send + Sync,
+{
+}
+
+#[async_trait::async_trait]
+pub trait CheckMemory<E> {
+    async fn check_memory(&self) -> anyhow::Result<Option<E>>;
+}
+
+#[async_trait::async_trait]
+impl<E, A> CheckMemory<E> for A
+where
+    A: db::Load<E> + std::ops::Deref<Target = SwapId>,
```

Comment:
This use of `Deref` is interesting because it only works based on `SwapId` being `Copy`, otherwise you couldn't go from the _reference_ you are getting from `Deref` to an owned instance.

I think `Deref` is not really supposed to be used like that :sweat_smile: 
It is mostly designed to make container types (`Mutex`, `Arc`, `Box`) work wherever a reference to the inner `T` is accepted.

An alternative design could just pass the `SwapId` as a parameter instead of referring to it implicitly through `self`?

Reply:
I hear you about the incorrect use of `Deref` :)

| An alternative design could just pass the SwapId as a parameter instead of referring to it implicitly through self?

I don't think this would work well because I would have to introduce the concept of a swap id in the `Do` trait, which doesn't need to know about it i.e. you could do an action for a swap without having an identifier for it.

Reply:
Would `Into` be more appropriate?